web/book/src/internals/modules.md

```diff
@@ -0,0 +1,95 @@
+# Modules
+
+PRQL needs modules that would:
+
+- allow importing declarations from other files,
+- have namespaces for things like `std`,
+- have hierarchical structure so we can represent files in directories,
+- have unambiguous module structure within a project,
+- be able to compile individual files that are a part of a project.
+
+## Basic definition
+
+Module is a namespace that contains declarations.
+
+Modules can be defined with mod keyword and a code block encased in curly
+braces:
+
+```
+mod my_module {
+    let bangers = (from tracks | take 10)
+}
+```
+
+## File loading
+
+If a module definition of `my_module` does not have a code block, it's contents
+are loaded from `./my_module.prql`.
+
+If the file does not exist, contents are loaded from file
+`./my_module/mod.prql`.
+
+When compilation is invoked compiler will search for the root of the module
+structure of the project:
+
+1. attempt to load `./project.prql`. If successful use it as the root,
+2. attempt to load `./mod.prql`. If successful visit parent directory and go
+   back to step 1.
+3. use compiled file as the root for the module structure.
+
+prql-compiler will provide an interface that has:
+
+- a function that takes a "file loader", which load files on-demand,
+- a simple function, that always report "file does not exist" to the compiler,
+  and can be used for compiling a single file to a single query.
```

Comment:
I would prefer that file loading was not customizable, that it always works in same predetermined predictable way.

Reply:
This is talking about interfaces prql-compiler will expose to different programs. For example, if you use it in some database connector (i.e. JDBC) there is no other files to load. Because only input a single source string, we need an interface that will report "file does not exist" when that source string tries to load some other file.

Will expand this section.